### PR TITLE
fix(treedb): quiesce dictionary work before replay cut

### LINKED ISSUE #3994: Stabilize authoritative-resource replay against late dictionary work
Parent execution: #3908
Blocks publication of: #3981 / PR #3993
Regression context: #3935 / PR #3936

## Goal

Make `TestPowerLossCertificationAuthoritativeResourcesPublicReopen` structurally deterministic at its terminal replay window: no asynchronous value-log dictionary work may publish or poison recovery after the witness arms the selected checkpoint cut.

## Why This Exists

Exact-head CI for PR #3993 exposed two failures in code untouched by that PR:

- Windows `windows-core-1`: the terminal checkpoint returned `collectionwal: recovery required: root publication cannot safely progress; reopen required` before reaching the intended cut.
- Linux `race-check (linux)-2`: a late value-log health update collided on `maindb/vlog_health.json.tmp.*`.

The failure is locally reproducible under the race detector. The witness currently treats 500 ms without observed durability events as quiescence, but the value-log dictionary trainer callback and periodic publisher can wake after that timing-only boundary. Once the replay window is armed, such work can consume the selected `AfterMetaSync` cut or poison the DB before the intended terminal checkpoint.

PR #3936 serialized concurrent dictionary profile publication and introduced the explicit replay window, but it did not make the terminal phase structurally free of future asynchronous dictionary work.

## Current Evidence

- Failing exact-head PR: #3993 at `95a0f4b3ca0391cd31a53d44786b94dd248ebc20`.
- GitHub failures:
  - https://github.com/snissn/gomap/actions/runs/30229933124/job/89866748574
  - https://github.com/snissn/gomap/actions/runs/30229933124/job/89866748566
- Local normal stress: `go test ./TreeDB -run '^TestPowerLossCertificationAuthoritativeResourcesPublicReopen$' -count=20` passed.
- Local race stress: `go test -race ./TreeDB -run '^TestPowerLossCertificationAuthoritativeResourcesPublicReopen$' -count=10` reproduced the premature recovery-required failure on repetition 9.

## Desired State

Use an explicit live-handle dictionary lifecycle barrier after asynchronous authoritative-resource setup:

1. keep the complete resource-creation persistence trace;
2. exclude concurrent writers/flushes while detaching every dictionary trainer;
3. close and wait for each trainer worker, then drain the last accepted-profile callback;
4. keep already-published dictionaries readable and permanently prevent trainer recreation on that handle;
5. only then begin the replay window and arm the deterministic terminal checkpoint cut.

This keeps the certification claim honest: it neither hides setup persistence events nor relabels a synthetic operation as a power-loss cut. It preserves the already-stabilized public DB storage epoch and removes the known future background writer through an explicit lifecycle fence instead of extending a sleep.

## Test-First Invariant

The existing race-stress failure is the red characterization: after the terminal replay window begins, the only eligible main-DB metadata sync must be the explicit boundary checkpoint. The implementation must make repeated normal and race runs select that boundary without background errors, premature recovery-required state, or duplicate temporary health files.

## Scope

- Add a caching-layer lifecycle method that permanently detaches dictionary trainers from one live DB handle, waits for worker shutdown, and drains in-flight publication.
- Prevent later trainer initialization on a quiesced handle while preserving access to the already published dictionary and dictionary-encoded frame.
- Expose the lifecycle seam to the external-package certificate only through a `_test.go` adapter.
- Preserve the full pre-window durability trace and post-cut public reopen assertions.
- Update comments/spec wording if the lifecycle boundary changes the documented witness contract.
- Add direct regression assertions that the terminal-phase options cannot restart dictionary training.

## Non-Goals

- Weakening or removing the power-loss witness.
- Ignoring or retrying a recovery-required result.
- Adding a cloud API, external-system comparison, or production control plane.
- Broad changes to dictionary compression, value-log format, or runtime checkpoint behavior.
- Treating the existing #3993 CI failures as waived because its feature files are unrelated.

## Performance Classification

**Not performance-relevant.** The intended change is confined to certification-test lifecycle setup and does not add work to production read, write, checkpoint, value-log, or query hot paths. No benchmark is required; stress, race, reopen, and latest-head CI evidence are authoritative.

## Execution And Merge Gates

- [ ] Capture the focused race failure or retain the exact existing red evidence above.
- [ ] Implement the smallest live-handle lifecycle barrier without changing the certificate's storage epoch.
- [ ] Pass the focused test repeatedly without `-race` and with `-race`.
- [ ] Pass affected TreeDB power-loss, dictionary, reopen, and package tests.
- [ ] Push one issue-scoped topic branch and open one PR closing this issue.
- [ ] Resolve mature AI review findings and obtain an exact-head Codex review result.
- [ ] Require latest-head CI green before merge; stale or unrelated-failure classification is not a waiver.
- [ ] Merge before rebasing and re-gating dependent PR #3993.

## Completion Criteria

- The authoritative-resource terminal phase has no timing-only dependency on dictionary publisher quiescence.
- Repeated race stress cannot reproduce premature recovery-required or value-log health temp-file collisions.
- The retained evidence still includes resource creation and uses the real public DB and checkpoint paths.
- Latest-head CI and exact-head mature review are clean.

### PR DESCRIPTION
**Pull request by snissn**
Closes #3994

Parent execution: #3908  
Blocks re-gating: #3981 / #3993

## Objective

Make the authoritative-resource power-loss certificate structurally deterministic at its terminal replay window. Before the witness arms the selected checkpoint cut, explicitly detach and drain all value-log dictionary trainers on the live DB handle and permanently prevent their recreation.

## Context

Exact-head CI for #3993 at `95a0f4b3ca0391cd31a53d44786b94dd248ebc20` exposed two failures in `TestPowerLossCertificationAuthoritativeResourcesPublicReopen`:

- Windows `windows-core-1`: the terminal checkpoint returned `collectionwal: recovery required: root publication cannot safely progress; reopen required` before reaching the intended cut.
- Linux `race-check (linux)-2`: a late value-log health update collided on `maindb/vlog_health.json.tmp.*`.

The race failure reproduced locally. The certificate's prior 500 ms quiet period was only a timing observation; dictionary trainer work accepted before the replay window could wake later, consume the selected `AfterMetaSync` cut, or poison the DB before the intended checkpoint.

This PR is the no-waiver prerequisite for rebasing and re-gating #3993.

## Non-goals

- Weakening, retrying, or bypassing the power-loss witness.
- Hiding setup persistence events or replacing the public checkpoint/reopen path.
- Reopening the DB to manufacture quiescence; that would create a new leaf-generation manifest epoch at the cut.
- Changing dictionary compression policy, value-log format, public TreeDB options, or ordinary checkpoint behavior.
- Treating #3993's failures as unrelated and merging around them.

## Design

- `compression.Trainer` now exposes `Wait`, backed by a worker-completion channel. `Close` stops acceptance; the queue worker drains accepted samples, waits for every training goroutine it spawned (including its `onAccept` callback), and only then completes `Wait`.
- `caching.DB.QuiesceValueLogDictTraining` serializes concurrent quiesce callers, follows checkpoint/close lock ordering (`flushMu`, then `writeMu`), permanently marks the live handle quiesced, detaches and deduplicates every trainer, closes them, waits after releasing the write/flush locks, and drains the final accepted-profile callback under `valueLogDictApplyMu`.
- Trainer creation refuses to restart after quiescence. Already-published dictionaries remain readable; only future sampling/training is disabled for that handle.
- Ordinary `DB.Close` also waits for trainer workers after closing them.
- The external-package certificate reaches the lifecycle seam through a `_test.go` adapter and invokes it before opening the terminal replay window.

Invariants:

1. No writer or flush can retain a trainer pointer while trainers are detached.
2. No trainer worker or accepted publication callback remains when quiescence returns.
3. Concurrent quiesce calls are safe and idempotent.
4. Quiescence does not close or reopen the DB and does not replace its storage epoch.
5. Published dictionary-backed frames stay decodable.

No on-disk format, public option, durability semantic, allocation length, mmap, or checksum behavior changes.

## Scope

- Includes:
  - `TreeDB/internal/compression/trainer.go`
  - `TreeDB/internal/compression/trainer_lifecycle_test.go`
  - `TreeDB/caching/db.go`
  - `TreeDB/caching/vlog_dict.go`
  - `TreeDB/caching/vlog_dict_quiesce_test.go`
  - authoritative-resource certificate and test-only adapter
- Excludes:
  - vector routing changes in #3993
  - compression format/policy changes
  - production callers of the explicit quiesce seam
  - cloud or external-system certification

## Correctness

Red characterization retained:

```text
go test -race ./TreeDB -run '^TestPowerLossCertificationAuthoritativeResourcesPublicReopen$' -count=10
```

Before this change, repetition 9 failed with premature `recovery required` after 108.302s. A normal 20-run stress passed, demonstrating the race-sensitive nature of the lapse.

Green evidence through exact head `352309a2d`:

```text
go test ./TreeDB/internal/compression -run '^TestTrainerWaitDrainsClosedWorker$' -count=100
go test ./TreeDB/internal/compression -run '^TestTrainerWait' -count=100
go test -race ./TreeDB/internal/compression -run '^TestTrainerWait' -count=100
go test -race ./TreeDB/caching -run '^TestQuiesceValueLogDictTrainingDetachesAndPreventsRestart$' -count=50
go test -race ./TreeDB -run '^TestPowerLossCertificationAuthoritativeResourcesPublicReopen$' -count=10
go test -race ./TreeDB -run '^TestPowerLossCertificationAuthoritativeResourcesPublicReopen$' -count=3
GOWORK=off go vet ./TreeDB/internal/compression ./TreeDB/caching ./TreeDB
GOWORK=off go test ./TreeDB/internal/compression ./TreeDB/caching ./TreeDB -count=1
git diff --check HEAD^
```

Results:

- trainer queue and spawned-job lifecycle: pass (100x normal and 100x race after the review fix)
- concurrent quiescence under race: pass (50x, 1.101s)
- authoritative-resource certificate under race: pass (10x, 111.816s), then pass again after final concurrency hardening (3x, 34.534s)
- affected-package vet: pass
- affected-package tests: pass (`compression` 0.061s, `caching` 90.327s, `TreeDB` 397.489s)
- diff check: clean

The trainer regression holds a modeled spawned training job open and proves `Wait` cannot return early. The direct quiescence test verifies detachment, completed worker shutdown, retained published dictionary state, no trainer restart, and eight concurrent callers.

## Performance

Not performance-relevant. The new synchronization is on explicit trainer quiescence and ordinary DB close; no read, write, checkpoint, value-log encode/decode, or query hot path gains per-operation work. The certificate is the only caller of the explicit lifecycle seam. No benchmark is claimed or required for this lifecycle correctness fix.

## Rollout / Toggle Plan

- No new option or default changes.
- No format migration.
- The certificate opts into quiescence through a test-only adapter.
- Revert this PR to remove the behavior; ordinary runtime operation otherwise remains unchanged apart from `Close` waiting for trainer shutdown.

## Follow-ups

- After merge, rebase #3993 onto the resulting `main`, rerun focused evidence, obtain a fresh exact-head Codex review, and require latest-head CI green.
- Continue #3908 with #3982 only after #3981 merges.

### AI Review Loop

- Codex latest-head review: clean for `352309a2d` ([result](https://github.com/snissn/gomap/pull/3995#issuecomment-5086681582)); retry also delivered clean
- Copilot latest-head review: pending if delivered
- CodeRabbit latest-head review: rate-limited after the final push; its earlier DRY-only thread was answered as intentionally distinct lock sequencing
- Review comments/threads resolved: yes; zero unresolved threads
- Re-requested after final meaningful push: yes (`@codex review` literal prompt)

---

### Scope

- [x] Single, clear objective
- [x] Explicit include and exclude lists
- [x] No unrelated feature reintroduction

### Safety / Correctness

- [x] Concurrency state and invariants are defined
- [x] No written-vs-durable `*Sync`, mmap, allocation-length, or checksum behavior changed
- [x] Deterministic lifecycle regressions added without sleep-based synchronization

### Tests

- [x] Focused regression and race tests pass repeatedly
- [x] Affected-package vet and package tests pass
- [x] Repository-wide platform/shard coverage completed through latest-head required CI

### Benchmarks / Measurements

- [x] Classified not performance-relevant with rationale; no hot-path benchmark applies

### Docs / Rollout

- [x] No format, config, or milestone documentation change applies
- [x] No production toggle is introduced

### Merge Gates

- [x] Exact-head mature Codex review clean
- [x] Latest-head required CI green (33/33)
- [x] Zero unresolved review threads

### CODE REVIEW

**Comment by snissn** [comment]
@codex review

File: TreeDB/internal/compression/trainer.go [hunk 1]
```diff
@@ -320,6 +322,15 @@ func (t *Trainer) Close() {
 	})
 }
 
+// Wait blocks until the trainer has drained every sample accepted before
+// Close and its background worker has stopped. Call Close before Wait.
+func (t *Trainer) Wait() {
+	if t == nil || t.doneCh == nil {
+		return
+	}
+	<-t.doneCh
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/compression/trainer.go:332]: Fixed in 352309a2d. The queue worker is now the sole producer of training-job WaitGroup additions; after sampleCh closes and drains, it waits for every spawned train goroutine before closing doneCh. This makes Add/Wait ordering structural and causes both Trainer.Wait and DB quiescence/Close to include in-flight training plus onAccept. A regression holds a modeled spawned job open and proves Wait cannot return early; it passed 100 normal and 100 race runs. Concurrent DB quiescence passed 50 race runs, and the authoritative-resource race certificate passed 3 more runs.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@codex review

File: TreeDB/caching/vlog_dict.go [hunk 2]
```diff
@@ -957,6 +960,63 @@ func (db *DB) ensureValueLogDictTrainer() {
...
+	for trainer := range closed {
+		trainer.Wait()
+	}
+
+	// A final accepted-profile callback may already be inside publication when
+	// the trainers are detached. Drain it before returning. Later periodic-loop
+	// attempts see no trainer and cannot mutate the dictionary store.
+	db.valueLogDictApplyMu.Lock()
+	db.valueLogDictApplyMu.Unlock()
+}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/caching/vlog_dict.go:1018]: Verified against exact head and keeping the two sequences local. They are similar mechanically but have intentionally different synchronization contracts: Quiesce marks the handle permanently quiesced and detaches/closes while holding flushMu+writeMu, must release those locks before waiting for trainers, then drains valueLogDictApplyMu; Close has already released flushMu+writeMu, waits for db.wg and drains valueLogDictApplyMu before trainer teardown, then continues destructive store/reader shutdown. A shared detach/close/wait helper callable under the write/flush locks would either wait under those locks (the deadlock risk this design explicitly avoids) or hide the required release boundary. This is a non-correctness DRY suggestion, not a lapse; no code change is warranted in this narrowly scoped lifecycle fix.
